### PR TITLE
[js] Add missing built-in APIs

### PR DESCRIPTION
**Pull request by terurou**

### CODE REVIEW

**Comment by nadako** [comment]
Wow, this looks really well-prepared, thank you for this! I'll review it when get time, or maybe @haxiomic or someone could take a look (also, maybe some of these should be generated from some IDL?)

**Comment by terurou** [comment]
I have no confidence about the added Object's methods. So, please review it carefully.

> maybe some of these should be generated from some IDL?

I think that `Intl` and `WebAssembly` should be generated from some IDL. But I couldn't find good (Web)IDL.

File: std/js/lib/Math.hx [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+		Returns the largest of zero or more numbers.
+	**/
+	@:overload(function(args:Rest<Float>):Float {})
+	@:pure static function max(args:Rest<Int>):Int;
+
+	/**
+		Returns the smallest of zero or more numbers.
+	**/
+	@:overload(function(args:Rest<Float>):Float {})
+	@:pure static function min(args:Rest<Int>):Int;
```
Comment (haxiomic) [inline std/js/lib/Math.hx:200]: Nice touch, thanks for adding separate signatures for Int and Floats :)!

**Comment by haxiomic** [comment]
Hey @terurou, this is a great contribution! Thank you :)!

I see what you mean about Object, but I think the approach you've taken is fine :). I like that you've added an abstract over the entries array.

I've had a run through and I can't see any wrong so LGTM from my perspective @nadako

File: std/js/lib/intl/Collator.hx [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	/**
+		Strings that differ in base letters, accents and other diacritic marks, or case compare as unequal.
+		Other differences may also be taken into consideration.
+		Examples: a ≠ b, a ≠ á, a ≠ A.
+	**/
+	var Variant = "variant";
+}
+
+enum abstract CaseFirst(String) {
+    var Upper = "upper";
```
Comment (nadako) [inline std/js/lib/intl/Collator.hx]: indentation is off here, but oh well :)

File: std/js/lib/intl/DateTimeFormat.hx [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+	var ?timeZoneName:TimeZoneName;
+}
+
+typedef DateTimeFormatResolvedOptions = {
+	/**
+		The BCP 47 language tag for the locale actually used.
+		If any Unicode extension values were requested in the input BCP 47 language tag that led to this locale,
+		the key-value pairs that were requested and are supported for this locale are included in `locale`.
+	**/
+	var locale(default,never):String;
```
Comment (nadako) [inline std/js/lib/intl/DateTimeFormat.hx]: shouldn't these be `final`?

File: std/js/lib/webassembly/Global.hx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+package js.lib.webassembly;
```
Comment (nadako) [inline std/js/lib/webassembly/Global.hx:22]: I wonder if webassembly stuff should be in just `js.webassembly`? @haxiomic any opinion?

**Comment by nadako** [comment]
Added a couple minor comments, but this looks really good!

File: std/js/lib/webassembly/Global.hx (hunk 4 above)
Comment (haxiomic) [reply std/js/lib/webassembly/Global.hx:22]: I don't have a strong opinion about this one, although I'm leaning to keeping in `lib` for consistency (the implication being `js.lib` = top-level native js externs and `js.*` = haxe-specific types that make working with the js target easier)

Comment (nadako) [reply std/js/lib/webassembly/Global.hx:22]: My line of thought is that `js.lib` contains standard APIs that are available in any JS engine, while platform-specific APIs reside in `js.*` sub-packages, just like `js.html` and `js.node`, that's why I thought it makes sense to also have `js.webassembly`.

Comment (haxiomic) [reply std/js/lib/webassembly/Global.hx:22]: I think WebAssembly (despite the name) _is_ supposed to be a standard JS engine API (it's available in v8 at least)
https://github.com/v8/v8/blob/master/src/wasm/wasm-js.cc#L2095
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly#Browser_compatibility

File: std/js/lib/intl/DateTimeFormat.hx (hunk 3 above)
Comment (terurou) [reply std/js/lib/intl/DateTimeFormat.hx]: Can I use `final`? There are yet many `(default, never)` style codes in other places.

Comment (Gama11) [reply std/js/lib/intl/DateTimeFormat.hx]: Codebases like hxnodejs still use the latter for backwards compatibility with Haxe 3.4. The std lib doesn't need to be backwards-compatible with older Haxe versions though.

Comment (terurou) [reply std/js/lib/intl/DateTimeFormat.hx]: Thanks. I'll change these to 'final'.

File: std/js/lib/webassembly/Global.hx (hunk 4 above)
Comment (terurou) [reply std/js/lib/webassembly/Global.hx:22]: WebAssembly is available in ChakraCore and node-chakuracore, too.
I think `js.lib` is better. (I think `js.webassembly` is also right.)

File: std/js/lib/intl/Collator.hx (hunk 2 above)
Comment (terurou) [reply std/js/lib/intl/Collator.hx]: Thanks! I'll fix it.

**Comment by nadako** [comment]
Thank you very much!

**Pull request merged**